services/src/main/java/org/keycloak/protocol/oidc/grants/OAuth2GrantTypeBase.java

```diff
@@ -0,0 +1,278 @@
+ * limitations under the License.
+ */
+
+package org.keycloak.protocol.oidc.grants;
+
+import jakarta.ws.rs.core.HttpHeaders;
+import jakarta.ws.rs.core.MediaType;
+import jakarta.ws.rs.core.MultivaluedMap;
+import jakarta.ws.rs.core.Response;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Function;
+
+import org.jboss.logging.Logger;
+
+import org.keycloak.Config;
+import org.keycloak.OAuth2Constants;
+import org.keycloak.OAuthErrorException;
+import org.keycloak.common.ClientConnection;
+import org.keycloak.common.Profile;
+import org.keycloak.common.VerificationException;
+import org.keycloak.constants.AdapterConstants;
+import org.keycloak.events.Errors;
+import org.keycloak.events.EventBuilder;
+import org.keycloak.http.HttpRequest;
+import org.keycloak.http.HttpResponse;
+import org.keycloak.models.AuthenticatedClientSessionModel;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.ClientSessionContext;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.KeycloakSessionFactory;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserModel;
+import org.keycloak.models.UserSessionModel;
+import org.keycloak.protocol.oidc.OIDCAdvancedConfigWrapper;
+import org.keycloak.protocol.oidc.TokenManager;
+import org.keycloak.protocol.oidc.utils.AuthorizeClientUtil;
+import org.keycloak.rar.AuthorizationRequestContext;
+import org.keycloak.representations.AccessToken;
+import org.keycloak.representations.AccessTokenResponse;
+import org.keycloak.representations.dpop.DPoP;
+import org.keycloak.services.CorsErrorResponseException;
+import org.keycloak.services.ServicesLogger;
+import org.keycloak.services.clientpolicy.ClientPolicyContext;
+import org.keycloak.services.clientpolicy.ClientPolicyException;
+import org.keycloak.services.resources.Cors;
+import org.keycloak.services.util.AuthorizationContextUtil;
+import org.keycloak.services.util.DPoPUtil;
+import org.keycloak.services.util.MtlsHoKTokenUtil;
+import org.keycloak.util.TokenUtil;
+
+/**
+ * Base class for OAuth 2.0 grant types
+ *
+ * @author <a href="mailto:demetrio@carretti.pro">Dmitry Telegin</a> (et al.)
+ */
+public abstract class OAuth2GrantTypeBase implements OAuth2GrantType {
+
+    private static final Logger logger = Logger.getLogger(OAuth2GrantTypeBase.class);
+
+    protected KeycloakSession session;
+    protected RealmModel realm;
+    protected ClientModel client;
+    protected OIDCAdvancedConfigWrapper clientConfig;
+    protected ClientConnection clientConnection;
+    protected Map<String, String> clientAuthAttributes;
+    protected MultivaluedMap<String, String> formParams;
+    protected EventBuilder event;
+    protected Cors cors;
+    protected TokenManager tokenManager;
+    protected DPoP dPoP;
+    protected HttpRequest request;
+    protected HttpResponse response;
+    protected HttpHeaders headers;
+
+    protected void initialize(Context context) {
```

Comment:
Do we need to call this method from every single implementation? Doesn't make more sense to have each implementation defining their fields (even if they need the same set)?

It will make the methods in this base class more self-contained?

Reply:
In regards to the fields in OAuth2GrantTypeBase, they all need to be there because the class contains common methods that depend on that fields.

But I agree about `initialize()` calls being redundant. I have replaced it with `setContext()` that will be called only once from `TokenEndpoint`.